src/stacked_borrows.rs

```diff
@@ -92,27 +92,90 @@ pub enum Permission {
     Disabled,
 }
 
-/// An item in the per-location borrow stack.
-#[derive(Copy, Clone, Hash, PartialEq, Eq)]
-pub struct Item {
-    /// The permission this item grants.
-    perm: Permission,
-    /// The pointers the permission is granted to.
-    tag: SbTag,
-    /// An optional protector, ensuring the item cannot get popped until `CallId` is over.
-    protector: Option<CallId>,
+impl Permission {
+    fn to_bits(self) -> u64 {
+        match self {
+            Permission::Unique => 0,
+            Permission::SharedReadWrite => 1,
+            Permission::SharedReadOnly => 2,
+            Permission::Disabled => 3,
+        }
+    }
+
+    fn from_bits(perm: u64) -> Self {
+        match perm {
+            0 => Permission::Unique,
+            1 => Permission::SharedReadWrite,
+            2 => Permission::SharedReadOnly,
+            3 => Permission::Disabled,
+            _ => unreachable!(),
+        }
+    }
 }
 
-impl fmt::Debug for Item {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "[{:?} for {:?}", self.perm, self.tag)?;
-        if let Some(call) = self.protector {
-            write!(f, " (call {})", call)?;
+mod item {
+    use super::{Permission, SbTag};
+    use std::fmt;
+
+    /// An item in the per-location borrow stack.
+    #[derive(Copy, Clone, Hash, PartialEq, Eq)]
+    pub struct Item(u64);
+
+    // An Item contains 3 bitfields:
+    // * Bits 0-61 store an SbTag
+    // * Bits 61-63 store a Permission
+    // * Bit 64 stores a flag which indicates if we have a protector
+    const TAG_MASK: u64 = u64::MAX >> 3;
+    const PERM_MASK: u64 = 0x3 << 61;
+    const PROTECTED_MASK: u64 = 0x1 << 63;
+
+    const PERM_SHIFT: u64 = 61;
+    const PROTECTED_SHIFT: u64 = 63;
+
+    impl Item {
+        pub fn new(tag: SbTag, perm: Permission, protected: bool) -> Self {
+            assert!(tag.0.get() <= TAG_MASK);
+            let packed_tag = tag.0.get();
+            let packed_perm = perm.to_bits() << PERM_SHIFT;
+            let packed_protected = (protected as u64) << PROTECTED_SHIFT;
+
+            let new = Self(packed_tag + packed_perm + packed_protected);
+
+            debug_assert!(new.tag() == tag);
+            debug_assert!(new.perm() == perm);
+            debug_assert!(new.protected() == protected);
+
+            new
+        }
+
+        /// The pointers the permission is granted to.
+        pub fn tag(self) -> SbTag {
+            unsafe { SbTag(std::num::NonZeroU64::new_unchecked(self.0 & TAG_MASK)) }
```

Comment:
Please add a SAFETY comment and a justification. We have *very little* `unsafe` in Miri, so if this does not make a measurable difference in a benchmark, I'd prefer to avoid the `unsafe` here.

Reply:
This was absolutely premature optimization on my part.